ebpf/tracer_test.go

```diff
@@ -559,6 +601,46 @@ func BenchmarkTCPSend(b *testing.B) {
 	runBenchtests(b, payloadSizesTCP, "eBPF", benchSendTCP)
 }
 
+func benchTCPClose(c func() (*Connections, error)) func(b *testing.B) {
+	collect := func() (*Connections, error) { return nil, nil }
+
+	if c != nil {
+		collect = c
+	}
+
+	echoOnMessage := func(c net.Conn) {
+		r := bufio.NewReader(c)
+		for {
+			buf, err := r.ReadBytes(byte('\n'))
+			if err == io.EOF {
+				c.Close()
+				return
+			}
+			c.Write(buf)
+		}
+	}
+
+	return func(b *testing.B) {
+		end := make(chan struct{})
+		server := NewTCPServer(echoOnMessage)
+		server.Run(end)
+		for i := 0; i < b.N; i++ {
+			c, err := net.DialTimeout("tcp", server.address, 5*time.Second)
+			if err != nil {
+				b.Fatal(err)
+			}
+
+			c.Write([]byte("hello"))
+			err = c.Close()
+			if err != nil {
+				b.Fatalf("can't close connection: %s", err)
+			}
+		}
+		collect()
```

Comment:
I think these benchmarks are not measuring the overhead of collection. We want the  `collect()` call inside the for loop as that's the part that's doing a lot of work.

```
func bench(b testing.B) {
   t := newTracer()
   // wait for t to initialize
   t.Run()
   go func() {
     server := NewTCPServer()
     server.Run(end)
     for { // some stopping condition
        // send message 
     }
   }
   b.Run("test collect", func(b) {
      for i := 0; i < b.N; i++ {
          collect()
       }
    }
}
```

That approach won't work for the perf buffer version because the work isn't being done by collect() in that version but in a callback. And much of the work is done in whatever goroutine kicked off by InitPerfMap. 

For a fair comparison between this and the perfbuffer version, we might want to run the tracer + nettop for a few minutes and sample CPU/memory instead of writing benchmarks.

Reply:
Yeah I added the collect as a third benchmark to take into account the 2 ways of retrieving the collections, I started to go the way of running the tracer and monitoring what happens at the system level, I will update the PR comments when I have results 👍